src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/clusterinteraction/ClusterDetailFragment.kt

```diff
@@ -49,45 +52,120 @@ class ClusterDetailFragment : Fragment() {
-      clusterAutoCompleteSetup(
+      if (requireArguments().getSerializable(HISTORY_COMMAND) != null) {
+        autoComplete(
+          historyCommand,
+          clusterAutoCompleteTv,
+          commandAutoCompleteTv,
+          parameterList,
+          inflater,
+          callbackList
+        )
+      } else {
+        commandAutoCompleteTv.visibility = View.GONE
+        clusterAutoCompleteSetup(
+          clusterAutoCompleteTv,
+          commandAutoCompleteTv,
+          parameterList,
+          callbackList
+        )
+        commandAutoCompleteSetup(commandAutoCompleteTv, inflater, parameterList, callbackList)
+      }
+      setInvokeCommandOnClickListener(
+        invokeCommand,
+        callbackList,
         clusterAutoCompleteTv,
         commandAutoCompleteTv,
-        parameterList,
-        callbackList
+        parameterList
       )
-      commandAutoCompleteSetup(commandAutoCompleteTv, inflater, parameterList, callbackList)
-      invokeCommand.setOnClickListener {
-        callbackList.removeAllViews()
-        val commandArguments = HashMap<String, Any>()
-        parameterList.forEach {
-          val type =
-            selectedInteractionInfo.commandParameters[it.clusterParameterNameTv.text.toString()]!!.type
-          val data = castStringToType(it.clusterParameterData.text.toString(), type)!!
+    }
+  }
 
-          commandArguments[it.clusterParameterNameTv.text.toString()] = data
-        }
-        selectedInteractionInfo.getCommandFunction()
-          .invokeCommand(selectedCluster, selectedCommandCallback, commandArguments)
+  private fun setInvokeCommandOnClickListener(
+    invokeCommand: Button,
+    callbackList: LinearLayout,
+    clusterAutoCompleteTv: AutoCompleteTextView,
+    commandAutoCompleteTv: AutoCompleteTextView,
+    parameterList: LinearLayout
+  ) {
+    invokeCommand.setOnClickListener {
+      callbackList.removeAllViews()
+      val commandArguments = HashMap<String, Any>()
+      clusterInteractionHistoryList.addFirst(
+        HistoryCommand(
+          clusterAutoCompleteTv.text.toString(),
+          commandAutoCompleteTv.text.toString(),
+          mutableListOf(),
+          null,
+          null
+        )
+      )
+      parameterList.forEach {
+        val parameterName = it.clusterParameterNameTv.text.toString()
+        val castType =
+          selectedInteractionInfo.commandParameters[parameterName]!!.type
+        val data = castStringToType(it.clusterParameterData.text.toString(), castType)!!
+        commandArguments[it.clusterParameterNameTv.text.toString()] = data
+        clusterInteractionHistoryList[0].parameterList.add(
+          HistoryParameterInfo(
+            parameterName,
+            data.toString(),
+            castType
+          )
+        )
       }
+      selectedInteractionInfo.getCommandFunction()
+        .invokeCommand(selectedCluster, selectedCommandCallback, commandArguments)
+    }
+  }
+
+  private fun autoComplete(
```

Comment:
Can you rename this to actually describe what the method does, specifically how it's for the history usecase?